server/src/domain/media/media.util.ts

```diff
@@ -0,0 +1,192 @@
+import { SystemConfigFFmpegDto } from '../system-config/dto';
+import { BitrateDistribution, TranscodeOptions, VideoCodecSWConfig, VideoStreamInfo } from './media.repository';
+
+abstract class BaseConfig implements VideoCodecSWConfig {
+  constructor(protected config: SystemConfigFFmpegDto) {}
+
+  getOptions(stream: VideoStreamInfo) {
+    const options = {
+      inputOptions: this.getBaseInputOptions(),
+      outputOptions: this.getBaseOutputOptions(),
+      twoPass: this.eligibleForTwoPass(),
+    } as TranscodeOptions;
+    const filters = this.getFilterOptions(stream);
+    if (filters.length > 0) {
+      options.outputOptions.push(`-vf ${filters.join(',')}`);
+    }
+    options.outputOptions.push(...this.getPresetOptions());
+    options.outputOptions.push(...this.getThreadOptions());
+    options.outputOptions.push(...this.getBitrateOptions());
+
+    return options;
+  }
+
+  getBaseInputOptions(): string[] {
+    return [];
+  }
+
+  getBaseOutputOptions() {
+    return [
+      `-vcodec ${this.config.targetVideoCodec}`,
+      `-acodec ${this.config.targetAudioCodec}`,
+      // Makes a second pass moving the moov atom to the beginning of
+      // the file for improved playback speed.
+      '-movflags faststart',
+      '-fps_mode passthrough',
+    ];
+  }
+
+  getFilterOptions(stream: VideoStreamInfo) {
+    const options = [];
+    if (this.shouldScale(stream)) {
+      options.push(`scale=${this.getScaling(stream)}`);
+    }
+
+    return options;
+  }
+
+  getPresetOptions() {
+    return [`-preset ${this.config.preset}`];
+  }
+
+  abstract getBitrateOptions(): Array<string>;
+  abstract getThreadOptions(): Array<string>;
```

Comment:
What are these abstract methods for?

Reply:
The bitrate and thread settings are more codec-specific so I didn't bother with a default implementation for them. But seeing as how H.264 and HEVC use the same settings for `getBitrateOptions`, it would be cleaner to just use that for the base class.